src/builtins/mod.rs

```diff
@@ -585,34 +585,29 @@ fn builtin_which(args: &[&str], shell: &mut Shell) -> i32 {
         return SUCCESS
     }
 
-    if args[1..].len() != 1 {
-        let stderr = io::stderr();
-        let mut stderr = stderr.lock();
-        let _ = stderr.write_all(b"which takes one argument\n");
-        return BAD_ARG;
-    }
-
-    let command = args[1];
-
-    if let Some(alias) = shell.variables.aliases.get(command) {
-        println!("{}: alias to {}", command, alias);
-        SUCCESS
-    } else if shell.builtins.contains_key(command) {
-        println!("{}: built-in shell command", command);
-        SUCCESS
-    } else if shell.functions.contains_key(command) {
-        println!("{}: function", command);
-        SUCCESS
-    } else {
-        for path in env::var("PATH").unwrap_or("/bin".to_string()).split(sys::PATH_SEPARATOR) {
-            let executable = Path::new(path).join(command);
-            if executable.is_file() {
-                println!("{}", executable.display());
-                return SUCCESS;
+    let mut result = SUCCESS;
+    'outer: for i in 1..args.len() {
+        let command = args[i];
+        if let Some(alias) = shell.variables.aliases.get(command) {
+            println!("{}: alias to {}", command, alias);
+            continue;
+        } else if shell.builtins.contains_key(command) {
+            println!("{}: built-in shell command", command);
+            continue;
+        } else if shell.functions.contains_key(command) {
```

Comment:
Should functions come before builtins?

Reply:
That was how it was before I edited it.

Reply:
May want to change it while you're at it.

Reply:
Sure.

Reply:
I believe you could also just write this as so:

```rust
'outer: for &command in &args[1..] { }
```

And then not have to manually dereference it,

Reply:
Thanks, sorry about the mistakes, I only recently began learning rust.

Reply:
It's fine. This is part of the purpose of submitting and reviewing pull requests -- pointing out tricks that you may not be aware of yet.